### PR TITLE
Explain HNT > DC conversion with widget

### PR DESCRIPTION
**Pull request by jthiller**
Better explain HNT to DC conversion and the oracle's role in the conversion price though the use of a widget instead of static examples.

While not a high-priority update, creating these sorts of experiences around token fundamentals help to guide critical understanding of Helium... but mostly this was a fun thing to build on a holiday evening.

Widget:
- directly accesses pyth oracle to show a realtime estimation
- allows user to drag slider to visualize oracle price impact on HNT issued per DC or DC burned per DC amount ($).
- gracefully falls back to defaults in case pyth price cannot be fetched.
- clicking on the live price updates the estimations.

Export-1700970010724

### CODE REVIEW

File: docs/tokens/data-credit.mdx [hunk 1]
```diff
@@ -20,14 +20,27 @@ For further reading on DC's place in the overall mechanics of the Helium network
 [burn and mint economics](/tokens/hnt-token/#burn-and-mint-economics) in the HNT token
 documentation.
 
+## Data Credit Fundamentals
+
+**One Data Credit equals $0.00001 USD.**
+
+The amount of DC produced by burning HNT will fluctuate based on the USD price of HNT as reported by
```
Comment (mbthiery) [inline docs/tokens/data-credit.mdx]: by burning `ONE` HNT. Basically trying to make this point more explicit.

File: src/theme/HntToDcSimulator.jsx [hunk 2]
```diff
@@ -0,0 +1,130 @@
+import React, { useState, useEffect } from "react";
+import { Connection, PublicKey } from "@solana/web3.js";
+import { PythConnection } from "@pythnetwork/client";
+
+export const HntToDcSimulator = () => {
+  const [liveHntPrice, setLiveHntPrice] = useState(0);
+  const [simulatedHntPrice, setSimulatedHntPrice] = useState(1);
+  const [dcAmount, setDcAmount] = useState(100000);
```
Comment (mbthiery) [inline src/theme/HntToDcSimulator.jsx]: this is derived from `simulatedHntPrice`, there is no need to hold it in state.

File: src/theme/HntToDcSimulator.jsx [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+        <div
+          style={{
+            display: "flex",
+            justifyContent: "space-between",
+            alignItems: "baseline",
+          }}
+        >
+          <p>Simulated HNT Oracle Price: ${simulatedHntPrice.toFixed(2)}</p>
+          {liveHntPrice ? (
+            <button
```
Comment (mbthiery) [inline src/theme/HntToDcSimulator.jsx]: I only knew to click on this by looking at the code. 
As it stands a user would only do this if they happened to hover over the text and put two and two together. I'd recommend we make this more explicit.

File: src/theme/HntToDcSimulator.jsx [hunk 4]
```diff
@@ -0,0 +1,130 @@
...
+              onClick={handleSetToLivePrice}
+              style={{
```
Comment (mbthiery) [inline src/theme/HntToDcSimulator.jsx]: nit: we could clean this up a tad by moving styles into `FILENAME.module.css`.
From there `import styles from "./FILENAME.module.css"` and then use `className={styles.name}`

File: src/theme/HntToDcSimulator.jsx [hunk 5]
```diff
@@ -0,0 +1,130 @@
+import React, { useState, useEffect } from "react";
+import { Connection, PublicKey } from "@solana/web3.js";
+import { PythConnection } from "@pythnetwork/client";
+
+export const HntToDcSimulator = () => {
+  const [liveHntPrice, setLiveHntPrice] = useState(0);
+  const [simulatedHntPrice, setSimulatedHntPrice] = useState(1);
+  const [dcAmount, setDcAmount] = useState(100000);
+  const [sliderRange, setSliderRange] = useState({ min: 0, max: 5 });
+  const [isInitialPriceSet, setIsInitialPriceSet] = useState(false);
```
Comment (mbthiery) [inline src/theme/HntToDcSimulator.jsx]: `isInitialPriceSet` can be derived from `liveHntPrice` since we can fairly assume that HNT will not reach 0 (or <.005). As such we can remove it from state.
For readability we can do `const isInitialPriceSet = !!liveHntPrice`

**Review by mbthiery — CHANGES_REQUESTED** [review]

File: src/theme/HntToDcSimulator.jsx [hunk 6]
```diff
@@ -0,0 +1,107 @@
+import { PythConnection } from '@pythnetwork/client'
+import { Connection, PublicKey } from '@solana/web3.js'
+import React, { useEffect, useState } from 'react'
+import styles from "./../css/HntToDcSimulator.module.css"
```
Comment (mbthiery) [inline src/theme/HntToDcSimulator.jsx]: nit - I'd collocate this at the same place as this jsx file. The styles will not apply to any other files so there is no need to have it in a shared /css folder.

File: src/theme/HntToDcSimulator.jsx [hunk 7]
```diff
@@ -0,0 +1,107 @@
...
+    pythConnection.start()
+    return () => pythConnection.stop()
+  }, [isInitialPriceSet])
+
+  const setInitialSliderValues = (price) => {
+    // based on queried price, update the initial position and min/max of the input
+    const roundedPrice = Math.round(price * 100) / 100
+    setSimulatedHntPrice(roundedPrice)
+    const upperLimit = Math.ceil(price / 5) * 5 + 5
+    let lowerLimit = Math.max(0, Math.floor(price / 5) * 5)
```
Comment (mbthiery) [inline src/theme/HntToDcSimulator.jsx]: nit: do we care to have lower limit increase with price? 
On the flip side we are doing 0.01 increments so that is a lot of steps... but its not like its hard to just drag. 🤷

File: src/theme/HntToDcSimulator.jsx [hunk 8]
```diff
@@ -0,0 +1,107 @@
...
+    setSliderRange({ min: lowerLimit, max: upperLimit })
+  }
+
+  const handleSliderChange = (event) => {
+    const newPrice = parseFloat(event.target.value)
+    setSimulatedHntPrice(newPrice)
+  }
+
+  const handleSetToLivePrice = () => {
+    const roundedPrice = Math.round(liveHntPrice * 100) / 100
```
Comment (mbthiery) [inline src/theme/HntToDcSimulator.jsx]: nit - This logic is used in two places and could be argued as non-trivial. We could extract this into a function then directly call
`setSimulatedHntPrice(getRoundedPrice(liveHntPrice))`

File: src/theme/HntToDcSimulator.jsx [hunk 9]
```diff
@@ -0,0 +1,107 @@
...
+            min={sliderRange.min}
+            max={sliderRange.max}
+            step="0.01"
+            value={simulatedHntPrice}
+            onChange={handleSliderChange}
+          />
+        </div>
+        <div className={styles.pricesContainer} >
+          <p>Simulated HNT Oracle Price: ${simulatedHntPrice.toFixed(2)}</p>
+          {liveHntPrice ? (
```
Comment (mbthiery) [inline src/theme/HntToDcSimulator.jsx]: nit - you can replace the ternary with `!!liveHntPrice && (<button>...</button>)`

File: src/theme/HntToDcSimulator.jsx [hunk 10]
```diff
@@ -0,0 +1,107 @@
...
+          </tr>
+        </thead>
+        <tbody>
+          <tr>
+            <td>{dcAmount.toLocaleString()} DC</td>
+            <td>{hntForTenDollars.toLocaleString()} HNT</td>
+          </tr>
+        </tbody>
+      </table>
+      <div className={`${styles.dcEstTable} ${styles.interactiveBox}`}>
```
Comment (mbthiery) [inline src/theme/HntToDcSimulator.jsx]: why is dcEstTable applied here. Pretty sure it isnt doing anything.

**Review by mbthiery — DISMISSED** [review]

File: src/theme/HntToDcSimulator.jsx (hunk 7 above)
Comment (jthiller) [reply src/theme/HntToDcSimulator.jsx]: Thinking here is if token is something like 50, the range becomes really wide for 0.01 increments. The +-10 sets an appropriate range based on fetched value at time of load.

The 0.01 increments feel a lot nicer with the slider than bigger increments.

File: src/theme/HntToDcSimulator.jsx (hunk 10 above)
Comment (jthiller) [reply src/theme/HntToDcSimulator.jsx]: It's lazy, but it's inheriting border-radius, box shadow, etc with a few overrides. Refactor?
image

Comment (mbthiery) [reply src/theme/HntToDcSimulator.jsx]: Ah got it.. missed that nuance. Why not do something like
```
.dcEstTable, .interactiveBox { 
shared-style1: value1;
shared-style2: value2;
}
.interactiveBox {
solo-style1: value1;
}
.dcEstTable {
solo-style1: value1;
}

```
By having them more explicitly shared/split helps avoid unintentional side-effects on a future refactor.

Comment (jthiller) [reply src/theme/HntToDcSimulator.jsx]: Ah yea I'm good with that. Limits the number of classes to look up too.

Comment (mbthiery) [reply src/theme/HntToDcSimulator.jsx]: yeah - exactly, when using css modules I generally like to keep it at 1 style per element except in a few edge cases to simplify things.

File: src/theme/HntToDcSimulator.jsx (hunk 7 above)
Comment (mbthiery) [reply src/theme/HntToDcSimulator.jsx]: Test it with a range of 0-100. It doesn't feel all that bad having such a large range. 
But if you want to keep it as is its also not a hill I'm looking to die on.

**Review by mbthiery — APPROVED** [review]
